design/one-pager-component-mutable-and-versioning.md

```diff
@@ -0,0 +1,475 @@
+            spec:
+              route:
+                - revision: frontend-a1234 # old revision
+                  percent: 60%
+                - revision: frontend-a75588698 # the new one
+                  percent: 40%
+```
+
+We use `[*]` to represent every array in the field.
+
+## A blue-green workflow case
+
+Let's say we have a `Rollout` Trait with definition like below:
+
+```yaml
+apiVersion: core.oam.dev/v1alpha2
+kind: TraitDefinition
+metadata:
+  name: rollouts.extend.oam.dev
+spec:
+  revision:
+    alwaysControl: true
+  definitionRef:
+    name: traffics.core.oam.dev
+```
+
+It was using in application with a component.
+
+```yaml
+apiVersion: core.oam.dev/v1alpha2
+kind: ApplicationConfiguration
+metadata:
+  name: example-appconfig
+spec:
+  components:
+    - componentName: frontend
+      traits:
+        - trait:
+            apiVersion: extend.oam.dev/v1
+            kind: Rollout
+            spec:
+              replica: 10
+              batch: 3
+              maxUnavailable: 1
+```
+
+Before changing the component, the App will be running in this stable state:
+
+```yaml
+apiVersion: core.oam.dev/v1alpha2
+kind: ContainerizedWorkload
+  metadata:
+    name: frontend-a75588698
+spec:
+  containers:
+    - name: my-cool-workload
+      image: example/very-cool-workload:0.1.2@sha256:verytrustworthyhash
+      cmd:
+        - "bash lscpu"
+---
+apiVersion: extend.oam.dev/v1
+kind: Rollout
+spec:
+  batch: 3
+  replica: 10
+  maxUnavailable: 1
+  workloadRef:
+    apiVersion: core.oam.dev/v1alpha2
+    kind: ContainerizedWorkload
+    name: frontend-a75588698
+status:
+  currentWorkloadRef:
+    apiVersion: core.oam.dev/v1alpha2
+    kind: ContainerizedWorkload
+    name: frontend-a75588698
+```
+
+When `frontend` component changed, new workload will be created, rollout trait will be updated,
+and the old workload will still be there.
+
+
+```yaml
+# the old one 
+apiVersion: core.oam.dev/v1alpha2
+kind: ContainerizedWorkload
+  metadata:
+    name: frontend-a75588698
+spec:
+  containers:
+    - name: my-cool-workload
+      image: example/very-cool-workload:0.1.2@sha256:verytrustworthyhash
+      cmd:
+        - "bash lscpu"
+---
+# the new one 
+apiVersion: core.oam.dev/v1alpha2
+kind: ContainerizedWorkload
+  metadata:
+    name: frontend-xas138742
+spec:
+  containers:
+    - name: my-cool-workload
+      image: example/very-cool-workload:new
+      cmd:
+        - "bash lscpu"
+---
+# rollout spec was pointing to the new workload, but status is still refer to the old.
+apiVersion: extend.oam.dev/v1
+kind: Rollout
+spec:
+  batch: 3
+  replica: 10
+  maxUnavailable: 1
+  workloadRef:
+    apiVersion: core.oam.dev/v1alpha2
+    kind: ContainerizedWorkload
+    name: frontend-xas138742
+status:
+  currentWorkloadRef:
+    apiVersion: core.oam.dev/v1alpha2
+    kind: ContainerizedWorkload
+    name: frontend-a75588698
+```
+
+Then the rollout trait will take control of the blue-green deploy progress.
```

Comment:
If the actor didn't specify the rollout trait who is responsible to control the rollout, by the workload itself?
How about the situation when the trait has been removed when doing rollout?

Reply:
> If the actor didn't specify the rollout trait who is responsible to control the rollout, by the workload itself?

Yes, the workload itself. We really have lots of workloads like Deployment/Statefulset/OpenKruise which have some rollout features.

> How about the situation when the trait has been removed when doing rollout?

First of all, that's an unexpected action, OAM runtime should prevent it happen. If this situation really happen, the AppConfig controller will update the workload and remove the old workload.